concordium-consensus/src/Concordium/External/DryRun.hs

```diff
@@ -0,0 +1,881 @@
+    (inQuota, remQuota) <- atomicModifyIORef' quotaRef $ \q ->
+        if q >= amount then (q - amount, (True, q - amount)) else (q, (False, q))
+    if inQuota then cont remQuota else return OutOfEnergyQuota
+
+-- | Load the state of a particular block in the dry-run session, and use its timestamp as the
+--  current timestamp for the session.
+dryRunLoadBlockState ::
+    -- | The dry-run handle.
+    StablePtr DryRunHandle ->
+    -- | Tag identifying the type of block hash input.
+    Word8 ->
+    -- | Payload data for the block hash input.
+    Ptr Word8 ->
+    -- | Vector in which to write the response.
+    Ptr ForeignVec ->
+    IO Int64
+dryRunLoadBlockState dryRunPtr bhiTag hashPtr outVec =
+    returnCode <$> do
+        DryRunHandle{..} <- deRefStablePtr dryRunPtr
+        tryTickQuota drhEnergyQuota costLoadBlockState $ \quotaRem -> do
+            input <- decodeBlockHashInput bhiTag hashPtr
+            res <-
+                runMVR
+                    ( liftSkovQueryBHIAndVersion
+                        ( \vc0 bp -> do
+                            drsTimestamp <- SkovV0.getSlotTimestamp $ SkovV0.blockSlot bp
+                            drsBlockState <- thawBlockState =<< blockState bp
+                            drs <- liftIO . newIORef $ DryRunState{..}
+                            return (drsTimestamp, EBlockStateContextV0 vc0 drs)
+                        )
+                        ( \vc1 bp _ -> do
+                            let drsTimestamp = SkovV1.blockTimestamp bp
+                            drsBlockState <- thawBlockState =<< blockState bp
+                            drs <- liftIO . newIORef $ DryRunState{..}
+                            return (drsTimestamp, EBlockStateContextV1 vc1 drs)
+                        )
+                        input
+                    )
+                    drhMVR
+            case res of
+                BQRNoBlock -> do
+                    writeProtoResponse (drhWriteToVector outVec) quotaRem DryRunErrorBlockNotFound
+                BQRBlock blkHash (ts, newBSC) -> do
+                    oldBlockStateContext <- atomicModifyIORef' drhBlockStateContext (Just newBSC,)
+                    forM_ oldBlockStateContext $ \oldBSC ->
+                        runWithEBlockStateContext drhMVR oldBSC $ \drs ->
+                            liftIO $ writeIORef drs (error "Dry run state dropped")
+                    writeProtoResponse
+                        (drhWriteToVector outVec)
+                        quotaRem
+                        ( DryRunSuccessBlockStateLoaded
+                            { drsBlockHash = blkHash,
+                              drsCurrentTimestamp = ts,
+                              drsProtocolVersion = bscProtocolVersion newBSC
+                            }
+                        )
+            return OK
+
+-- | A dummy 'StateHash' value that can be used to create a 'HashedPersistentBlockState' where the
+-- actual hash does not matter.
```

Comment:
There should be some explanation about the context and the need for this. What problem does it solve.